### ISSUE TITLE
Rule Change: no-use-before-define option to ignore top-level variables

### ISSUE THREAD

**Issue by barryam3** [eslint/eslint#19968]
### What rule do you want to change?

no-use-before-define

### What change do you want to make?

Generate fewer warnings

### How do you think the change should be implemented?

A new option

### Example code

```js
/*eslint no-use-before-define: ["error"]*/
export const foo = () => {
  topLevelVariable();
  [1].forEach(() => {
    otherVariable();
  });
  const otherVariable = () => {};
};
const topLevelVariable = () => {};
```

### What does the rule currently do for this code?

Error on `topLevelVariable` and `otherVariable`

### What will the rule do after it's changed?

Error on only `otherVariable`

### Participation

- [x] I am willing to submit a pull request to implement this change.

### Additional comments

Rationale: The `variables: true` option is required to catch bugs. For example, the following code is an error only with that option enabled:

```js
a.forEach((el, i) => {
  // error with `variables: true`
  // no error with `variables: false` since we're in an inner scope
  m.set(el, i);
});

const m = new Map();
```

I've seen bugs like this and want to enable `variables: true` however this would lead to thousands of errors in my codebase, since it's a common pattern for people to declare top-level variables, particularly arrow functions, out of order. Calling a function or otherwise using a variable at the top level of a file upon import rather than upon calling an imported function is rare, so these references are almost always safe.

I think there is reason to add a new option `variables: 'ignoreTopLevel'` which allows out-of-order references to variables that are declared at the top level only.

This issue is similar to #12986 but the outcome of that was to clarify that the `variables` option applies to arrow functions, and it's similar to #12473 but the recommendation from that was to restructure code or disable `functions`.

**Comment by nzakas** [comment]
Here are summaries and resolutions for the related issues mentioned:

**Issue #12986: no-use-before-define needs exception option for arrow function definitions present in common code patterns**
- Summary: This issue requested a new option for the `no-use-before-define` rule to ignore arrow function definitions, arguing that modern codebases (especially React and ES6 modules) often use interconnected arrow functions that are not evaluated until after all definitions are parsed. The current rule treats arrow functions like variables, leading to unnecessary warnings and forcing developers to refactor code or abandon modern syntax.
- Resolution: The team clarified that the `variables` option applies to arrow functions, but did not add a dedicated exception for them. The issue was closed with the recommendation to use existing options or refactor code as needed.

**Issue #12473: [no-use-before-define] False positive for mutually recursive functions**
- Summary: This issue reported that the rule flags mutually recursive function definitions (e.g., `isOdd` and `isEven`) as errors, even though the code is valid and works as intended. The user expected no error, but ESLint reported a use-before-define error for one of the functions.
- Resolution: The team determined that the rule was working as intended and recommended restructuring code or disabling the rule for such cases. No change was made to the rule, and the issue was closed.
---
*This summary was AI-generated.*

**Comment by nzakas** [comment]
Thanks for the suggestion. I understand what you're trying to do, I'm just not sure I agree with the proposal to exempt top-level variables from the rule. That seems like a small edge case when the larger issue is that arrow functions are lazily evaluated so any variable declared later in source order would be flagged. As a simple example:

```js
{
  a.forEach((el, i) => {
    // error with `variables: true`
    // no error with `variables: false` since we're in an inner scope
    m.set(el, i);
  });

  const m = new Map();
}
```

@eslint/eslint-team what do you think?

**Comment by JoshuaKGoldberg** [comment]
Agreed, I don't see a strong reason to exempt specifically top-level variables. +1 to the the reasoning+counter-example as posted.

**Comment by kirkwaiblinger** [comment]
Why are you not better off converting the arrow functions to `function`s, which are genuinely safe at runtime?

```js
const a = () => { b(); }

a(); // Runtime Error: Cannot access 'b' before initialization

const b = () => { console.log('hello world'); };
```

```js
const a = () => { b(); }

a(); // Runtime Error: 'b' is not a function

var b = () => { console.log('hello world'); };
```

```js
function a() {
  b();
}

a(); // logs 'hello world'

function b() { 
  console.log('hello world');
};
```

**Comment by barryam3** [comment]
> Why are you not better off converting the arrow functions to functions, which are genuinely safe at runtime?

We use React with TypeScript. If you want to ensure components conform to the `React.FC` interface you need to use an arrow function.

```tsx
const MyComponent: React.FC<MyProps> = props => {
  return <div>TODO</div>;
};
```

Additionally if you want to use a higher-order component like `React.memo` then you need a `const` declaration.

```jsx
const MyComponent = React.memo(props => {
  return <div>TODO</div>;
});
```

What I'm hearing from my team is we'd like to enable `variables: true` to prevent bugs like the one in your example.

> ```js
>   a.forEach((el, i) => {
>     // error with `variables: true`
>     // no error with `variables: false` since we're in an inner scope
>     m.set(el, i);
>   });
> 
>   const m = new Map();
> ```

But we want to be able to define top-level arrow functions in any order and are not worried about arrow functions being called out of order in the top level scope, since it's a module-based web app. This seemed like a use case that would be typical of TypeScript React codebases.

**Comment by kirkwaiblinger** [comment]
Gotcha... Same sort of deal as https://github.com/typescript-eslint/typescript-eslint/issues/10127 then 👍 

EDIT: Oh, which did eventually get incorporated into core (https://github.com/eslint/eslint/issues/19719) 🤔

**Comment by JoshuaKGoldberg** [comment]
> If you want to ensure components conform to the `React.FC` interface you need to use an arrow function.

[You can always have a function expression with `React.FC`](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQNwBQdWAHpLHMAHYxZQF5ZwACsTDoA3nThS4qYAC8sALnKoQyADbqycAD7l1yKAHMstOgF8GzVvFwQOqeAFEmycOqWYqMAHQAxAGEAHmEIUQA+OABeOAIAVw58YHs4AAowEVQASjgJaThKGDioDjgggBNgADc4XANUVAA5NywosQyw1B9ZBXNwnwGggHpKqvD6cyA):

```tsx
import * as React from 'react';

export interface Props {
    size: 'small' | 'large';
}

export const Example: React.FC<Props> = function (props) {
    return <div className={props.size}>...</div>;
}
```

I think your point stands on not having to use a _variable_ though. https://github.com/microsoft/TypeScript/issues/51556 is still open for using `satisfies` with a function declaration.

**Comment by nzakas** [comment]
We don't have agreement on the team to move forward with this, so closing.
